From 2523a49162adb0348745b29869797eefd72a8ced Mon Sep 17 00:00:00 2001
From: "ji, zhenlong z" <zhenlong.z.ji@intel.com>
Date: Mon, 20 Jan 2020 09:57:08 +0800
Subject: [PATCH] Ensure android is using the LSM module of SELinux

It may be better to do this thing in android-entry.

If we enable LSM stacking feature in the kernel, host and
containers can use different LSM modules. Such as, host runs
with apparmor enabled and android containers use selinux.

The proc attr of display is used to set a process's LSM to a
specific value. The process's LSM will inherit its parent by
default, so we need to set init's LSM to selinux explicitly.

Change-Id: Ie0ccc476b357faa10cc0a85de99f72da2a477468
Tracked-On:
Signed-off-by: ji, zhenlong z <zhenlong.z.ji@intel.com>
---
 init/init.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/init/init.cpp b/init/init.cpp
index 4fe115e92..95afd7230 100644
--- a/init/init.cpp
+++ b/init/init.cpp
@@ -634,6 +634,15 @@ int main(int argc, char** argv) {
         uint64_t start_ms = start_time.time_since_epoch().count() / kNanosecondsPerMillisecond;
         setenv("INIT_STARTED_AT", std::to_string(start_ms).c_str(), 1);
 
+        int display_fd = open("/proc/self/attr/display", O_WRONLY | O_CLOEXEC);
+        if (display_fd < 0) {
+            LOG(INFO) << "open init display attr failed ...";
+        } else {
+            if (write(display_fd, "selinux", 7) < 0) {
+                LOG(INFO) << "write init display attr failed ...";
+            }
+        }
+
         char* path = argv[0];
         char* args[] = { path, nullptr };
         execv(path, args);
-- 
2.20.1

